### PR TITLE
Enable strict mode in roslyn compiler

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/AspNetCore/issues/3645

### CODE REVIEW

File: src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs [hunk 1]
```diff
@@ -131,11 +131,6 @@ private class TestModelBinder : IModelBinder
             public Task BindModelAsync(ModelBindingContext bindingContext)
             {
                 var val = bindingContext.ValueProvider.GetValue(bindingContext.ModelName);
-                if (val == null)
-                {
-                    return Task.CompletedTask;
-                }
-
```
Comment (rynowak) [inline src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: This is a semantic change.

Comment (pakrym) [reply src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: ValueProviderResult is a struct for result would always be false

**Review by Tratcher — APPROVED** [review]

File: src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: This is probably a bug then? /cc @dougbu

Comment (dougbu) [reply src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: @rynowak what exactly do you want my thoughts on?

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public StringValues HeaderCacheControl
         {
             get
             {
-                StringValues value;
+                StringValues value = default;
```
Comment (dougbu) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs]: This is a generated file but I don't see changes in the generator. Suggest there should be.

Comment (pakrym) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs]: Generator change is in `src/Servers/Kestrel/tools/CodeGenerator/KnownHeaders.cs`

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs [hunk 3]
```diff
@@ -448,17 +448,12 @@ protected virtual void CopyToModel(object target, IEnumerable<TElement> sourceCo
 
         private static IEnumerable<string> GetIndexNamesFromValueProviderResult(ValueProviderResult valueProviderResult)
         {
-            IEnumerable<string> indexNames = null;
-            if (valueProviderResult != null)
+            var indexes = (string[])valueProviderResult;
+            if (indexes != null && indexes.Length > 0)
```
Comment (halter73) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: Isn't it impossible for `indexes` to be null here? `valueProviderResult` is a struct and the `(string[])` must cast either succeed or throw.

File: src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs (hunk 1 above)
Comment (rynowak) [reply src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: oh nvmt his is in test code. Just ignore then

Comment (halter73) [reply src/Mvc/test/WebSites/BasicWebSite/Controllers/ContactApiController.cs]: Is it OK to remove the if condition entirely (which doesn't change the current behavior) or should be replaced with something like `if (val.Length > 0)`?

**Review by halter73 — APPROVED** [review]

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs (hunk 3 above)
Comment (pakrym) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: Might be `null` for `default(StringValues)`

File: Directory.Build.props [hunk 4]
```diff
@@ -22,6 +22,8 @@
     <NoWarn>$(NoWarn);NU5125</NoWarn>
     <!-- Suppress warnings about using SemVer 2.0. -->
     <NoWarn>$(NoWarn);NU5105</NoWarn>
+    <!-- Enables Strict mode for Roslyn compiler -->
+    <Features>strict</Features>
```
Comment (natemcmaster) [inline Directory.Build.props]: Nit: can you put this in the propertygroup below that starts with the comment "Compilation options"?

**Review by natemcmaster — APPROVED** [review]

File: src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs (hunk 3 above)
Comment (halter73) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: Yowzers! Return null from an implicit conversion sounds pretty evil. We couldn't even return a static empty array?

Comment (benaadams) [inline src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: Could one line it?
```csharp
return (indexes == null || indexes.Length == 0) ? null : indexes;
```

Comment (Tratcher) [reply src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/CollectionModelBinder.cs]: No, that's expected to preserve transitivity. E.g. (string[])(StringValues)null;

**Pull request merged**